src/baal/modelwrapper.py

```diff
@@ -335,17 +337,39 @@ def predict_on_batch(self, data, iterations=1, cuda=False):
         Returns:
             Tensor, the loss computed from the criterion.
                     shape = {batch_size, nclass, n_iteration}
+
+        Raises:
+            raises RuntimeError if CUDA rans out of memory during data replication.
         """
         with torch.no_grad():
             if cuda:
                 data = data.cuda()
-            input_shape = data.size()
-            batch_size = input_shape[0]
-            data = torch.stack([data] * iterations)
-            data = data.view(batch_size * iterations, *input_shape[1:])
-            out = self.model(data)
-            out = map_on_tensor(lambda o: o.view([iterations, batch_size, *o.size()[1:]]), out)
-            out = map_on_tensor(lambda o: o.permute(1, 2, *range(3, o.ndimension()), 0), out)
+            if self.replicate_in_memory:
+                input_shape = data.size()
+                batch_size = input_shape[0]
+                try:
+                    data = torch.stack([data] * iterations)
+                except Exception as e:
+                    raise RuntimeError(
+                        '''CUDA ran out of memory while BaaL tried to replicate data. See the exception above.
+                    Use `replicate_in_memory=False` in order to reduce the memory requirements.
+                    Note that there will be some speed trade-offs''') from e
+                data = data.view(batch_size * iterations, *input_shape[1:])
+                try:
+                    out = self.model(data)
+                except Exception as e:
+                    raise RuntimeError(
+                        '''CUDA ran out of memory while BaaL tried to replicate data. See the exception above.
+                    Use `replicate_in_memory=False` in order to reduce the memory requirements.
+                    Note that there will be some speed trade-offs''') from e
+                out = map_on_tensor(lambda o: o.view([iterations, batch_size, *o.size()[1:]]), out)
+                out = map_on_tensor(lambda o: o.permute(1, 2, *range(3, o.ndimension()), 0), out)
+            else:
+                out = [self.model(data) for _ in range(iterations)]
+                if isinstance(out[0], Sequence):
```

Comment:
Couldn't we use `map_on_tensor` for this? this is just for unifying the code

Reply:
We're dealing with a lis of lists of tensors here, I would have to modify the `map_on_tensor` to account for this. If that's preferable I definitely can do it.

Reply:
well in that case, it is absolutely a design choice and not important in my opinion. We can keep it as it is. Perfect